### ISSUE TITLE
How much faster can we make `require`?

### ISSUE THREAD

**Issue by nathansobo** [electron/electron#169]
One thing we're really struggling with is the overhead for requiring packages upon starting Atom. Are we sure that `require` is as fast as it can possibly be? I'm assuming the answer is probably yes, but I thought it would be worth it to ask / investigate considering the slowness we originally found.

**Comment by yhwh** [comment]
@hachre I bet it's what http://enclosejs.com/ does.

**Comment by yangguo-chromium-org** [comment]
Implementer of the serializer/deserializer in V8 here.

Code caching and custom heap snapshot are both features that have recently been completed.

As someone already mentioned, V8 already been using startup snapshot to speed up startup for years. Though the serializer/deserializer had a lot of of limitations that prevent general use. Those have been fixed recently.

Code caching serializes the compiled JIT code of a top level script. When compiling a script via v8::ScripCompiler::Compile, you can request cache data. In a later call to it, you can provide that cache data. Usually deserializing is anywhere between 10 to 100 faster than paraing/compiling, depending on the script and code to be cached.

Then there's the custom startup snapshot. You can create a snapshot data blob via v8::V8::CreateSnapshotDataBlob, and pass a script to run before producing the snapshot. You can then pass that snapshot to v8::Isolate::New to recreate the Isolate captured in the snapshot. However, since calls to C++ functions cannot be replayed, things like calls to require won't work. You could work around this particular issue by inlining the required scripts, but that's not really elegant. If your startup code takes a long time to run, this is a good way to avoid repeating the same execution by simply caching the result.

**Comment by zcbenz** [comment]
@yangguo-chromium-org Thanks for the information!

> Code caching serializes the compiled JIT code of a top level script. When compiling a script via v8::ScripCompiler::Compile, you can request cache data. In a later call to it, you can provide that cache data. Usually deserializing is anywhere between 10 to 100 faster than paraing/compiling, depending on the script and code to be cached.

The last time (a few years ago) I looked into the compile cache I thought it only cached AST, it is amazing to know it can store compiled JIT code now.

I'm not an expert on this, but is (or will) it possible to compile a script with only the compiled JIT code? It would be great for source code protection and performance if we don't need to provide the actual JavaScript source code.

**Comment by yangguo-chromium-org** [comment]
@zcbenz: source protection is not an objective for v8, so it won't be supported out of the box. The source code is required for lazily compiled inner functions and for the optimizing compiler. If you disable both, the source code is not actually necessary, aside from some minor details that are easy to fix. AFAIK this is essentially what nw.js does.

**Comment by nathansobo** [comment]
@zcbenz What if we separated the DOM and native interactions into a second phase? Presumably `require` isn't an absolute requirement because we could concatenate the core scripts manually, or do something like `asar`... or am I missing something there?

**Comment by zcbenz** [comment]
> What if we separated the DOM and native interactions into a second phase?

@nathansobo We don't have to do that, my idea of using snapshot feature is like this:

1) Iterate all the JavaScript files and concatenate them into a single `snapshot_cache.js`:

```javascript
var __snapshotCache = {
  'src/atom.js': function(exports, require, module, __filename, __dirname) {
    ...
  },
  'src/buffered-node-process.js': function(exports, require, module, __filename, __dirname) {
    ...
  },
  '...': function(exports, require, module, __filename, __dirname) {
    ...
  },
};
```

2) Make a snapshot file from the concatenated file:

```bash
$ mksnapshot snapshot_cache.js atom.snapshot
```

Since module's code are only compiled and not actually called, we won't have the problem of native functions not working in snapshot.

3) Load the snapshot file in atom-shell, the `__snapshotCache` defined in  `snapshot_cache.js` will be available in window context:

```javascript
window.__snapshotCache = { ... };
```

4) Monkey-patch the `require` function to make it read module source code from `window.__snapshotCache` instead of reading files and compiling strings.

And we can go further by generating snapshot file when creating `asar` archive, and make atom-shell recognize snapshot file in `asar` archive, so users can get the speed for free. This can be quite complicated and we need to wait for at least Chrome 41.

**Comment by yangguo-chromium-org** [comment]
Note that inner functions are usually compiled lazily. That means that those that are not called when creating the snapshot don't actually get their code compiled and included in the snapshot. That may be a good thing, or not, depending on whether you actually call it later. Of course you can also disable lazy compilation via --nolazy.

I actually think the two-phase approach is better: start up atom as far as you can without calling the DOM, create the snapshot, and then use that as basis for the second phase.

If you just want to save compile time and have the compiled code ready, you might as well use the code cache instead of a full heap snapshot. The advantage of the heap snapshot is that you get not only compiled code without spending time compiling, but also the entire heap state without executing code to set it up.

**Comment by zcbenz** [comment]
@yangguo-chromium-org It is really helpful information, thanks a lot!

And I have a question on using external snapshot file: Chromium is already shipping a `snapshot_blob.bin`, and it seems that V8 only supports one snapshot file, so does it matter if we replace Chromium's snapshot file with Atom's one? Or should we find out a way to combine Chromium's snapshot file with Atom's one?

**Comment by yangguo-chromium-org** [comment]
@zcbenz I recently landed this: https://codereview.chromium.org/949623006/

With that, you can load an arbitrary snapshot blob into the memory and pass it via v8::Isolate::CreateParams when calling v8::Isolate::New. It will then use that snapshot to initialize the heap and as template to create each new context.

But maybe that's not even necessary. Afaik Chromium's snapshot file is a vanilla startup snapshot. So a custom snapshot is based on it either way.

Btw. about that two-phase approach: you should include the function that does the DOM set up in the snapshot by defining the function, and calling it inside a try-catch. By referencing the DOM, you will get a ReferenceError that is caught, but the unsuccessful call will at least have caused the function to compile and end up in the startup snapshot.

**Comment by yangguo-chromium-org** [comment]
Maybe to clarify: you can only use one snapshot to create an Isolate from. There is no way to use two isolates, if that's what you mean by having the Chromium snapshot and an additonal Atom snapshot. I'm not familiar with how Atom uses V8. If you have separate Isolates for Chromium and for Atom, then having two snapshots is the way to go. If you share one Isolate for both, then you probably want a snapshot that captures a state as far into starting up Atom as possible to speed up start up.

**Comment by yangguo-chromium-org** [comment]
I made a blog post for this. Maybe it helps.
http://www.hashseed.net/2015/03/improving-v8s-performance-using.html

**Comment by zcbenz** [comment]
@yangguo-chromium-org It is awesome to have an updated blog on improving V8's performance, your perspective will help us a lot!

**Comment by yhwh** [comment]
Hey guys just in case you missed it,
 nwjs has a new snapshotting tool called nwjc
https://github.com/nwjs/v8/blob/nw12/src/nwjc.cc
It removes code size limitations of the nwsnapshot.
https://github.com/nwjs/nw.js/wiki/Protect-JavaScript-source-code-with-v8-snapshot
Interesting to see their approach with evalNWBin()

**Comment by yangguo-chromium-org** [comment]
It's based on the same API described in my blog post. I've talked to @rogerwang about nwjs before.

**Comment by YuriSolovyov** [comment]
on asar part:
since asar package header contains all filenames with their paths, Model._resolveFilename could be almost and instant, right? since you don't have to traverse filesystem, but only load data by offset

**Comment by zcbenz** [comment]
@yangguo-chromium-org 

I'm experimenting with the snapshot feature with Chrome 41 (V8 version v4.1.0.21), [this patch](https://chromium.googlesource.com/v8/v8/+/39cd762c6c1675304dc9a3ba9637ac405a6eb211%5E%21/) has been manually applied to enable embedding custom script.

I generated the snapshot file with following command:

```bash
$ mksnapshot output.cc --startup_blob snapshot_blob.bin jquery.js
```

And then I overwrite atom-shell's `snapshot_blob.bin` with the one generated by `mksnapshot`.

With simple scripts this works fine, I can see global variables correctly set. But after trying with big scripts, for example a big function that wraps the source code of jQuery, the `snapshot_blob.bin` is generated without errors, but when loading it in atom-shell I got following crash:

```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000227

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libchromiumcontent.dylib      	0x000000010fa49c44 v8::Testing::DeoptimizeAll() + 1488788
1   libchromiumcontent.dylib      	0x000000010fa4c4c3 v8::Testing::DeoptimizeAll() + 1499155
2   libchromiumcontent.dylib      	0x000000010fa4d359 v8::Testing::DeoptimizeAll() + 1502889
3   libchromiumcontent.dylib      	0x000000010fc50a04 v8::Testing::DeoptimizeAll() + 3614036
4   libchromiumcontent.dylib      	0x000000010fb4af07 v8::Testing::DeoptimizeAll() + 2542167
5   libchromiumcontent.dylib      	0x000000010fc5a795 v8::Testing::DeoptimizeAll() + 3654373
6   libchromiumcontent.dylib      	0x000000010f8c59be v8::Isolate::New(v8::Isolate::CreateParams const&) + 174
7   libchromiumcontent.dylib      	0x000000011084a392 gin::IsolateHolder::IsolateHolder() + 242
8   libchromiumcontent.dylib      	0x000000010e302084 net::ProxyResolverV8::EnsureIsolateCreated() + 52
```

It seems that the snapshot feature is still not able to load big objects? Or is there a change that required to make it work not available in Chrome 41?

**Comment by yangguo-chromium-org** [comment]
I'm pretty sure I fixed a bunch of smaller issues since 41. Also, please do not run from a snapshot when creating one.

Embedding jquery.js won't work, since embedding a script requires running it. What the snapshot captures is the heap state after the script has executed. Jquery accesses the window object, which mksnapshot does not offer.

What you probably want instead is code caching, since it only compiles the code, but doesn't run it. Alternatively, you would have to refactor your code so that the snapshotted part does not access the window object.

**Comment by zcbenz** [comment]
What I did is wrapping jQuery's source code inside a function like this:

```javascript
var a = function() {
  // jQuery source code
};
```

So it will not actually run jQuery.

I'll have a retry when Chrome 42 releases, thanks!

**Comment by steelbrain** [comment]
Bump.

**Comment by steelbrain** [comment]
Looks like io.js has been working on this https://github.com/nodejs/io.js/commit/b14fd1a7202cd8001d1f6aed848445bad99ee15c  https://github.com/nodejs/io.js/commit/1bbf8d0720a061e281718f9e6a542da4268160b8

**Comment by subesokun** [comment]
Sorry for the naive question, but would it be possible to snapshot also the heap of the BrowserWindow? I'd be interested in how to speed up the initial loading of the content of the BrowserWindow and it'd be awesome if I can snapshot also the initial heap of that window after loading all JS files. If that's possible, how could I create & load such snapshot?

**Comment by yangguo-chromium-org** [comment]
I'm not familiar with atom itself. If BrowserWindow is a pure Javascript object, then this would work. If it however contain references to C++ for example via function templates, then no.

**Comment by subesokun** [comment]
@yangguo-chromium-org Thanks for the quick feedback. Is there actually some good documentation regarding the snapshot feature?

I've played now a bit around with the `mksnapshot` tool. Initially I thought that electron builds already its own `snapshot_blob.bin` but so far this file gets only copied out of the `libchromiumcontent/static_library` folder. By just calling `mksnapshot output.cc --startup_blob snapshot_blob.bin` it creates a snapshot with a similar size as the one contained in the `static_library` folder so I wonder if this by default shipped snapshot is just an "empty" snapshot that gets created via the command mentioned above? 
Then I've tried to add a large JS file (1.1 MB) like `mksnapshot output.cc --startup_blob snapshot_blob.bin build.js` but it ended in a segmentation fault :( With some small JS file it completed without problems.
```
#
# Fatal error in ../../v8/src/snapshot/mksnapshot.cc, line 163
# Check failed: blob.data.
#

==== C stack trace ===============================

 1: 0x9ffe8f
 2: 0x44c136
 3: __libc_start_main
 4: 0x44bd11
Illegal instruction (core dumped)
```
Basically if I'm able to include my JS file into this `snapshot_blob.bin` it will be loaded during initialization of the V8 or? But does that also mean that this heap will be available within the BrowserWindow context or will it have its own separated heap as a new window gets created...?

**Comment by subesokun** [comment]
Oh wow it actually works! I've added React.js to the snapshot and even without loading the library I can access it in the BrowserWindow :)

**Comment by yangguo-chromium-org** [comment]
mksnapshot produces a default snapshot if you don't provide any additional file, which is the usual snapshot what v8 uses. If you pass a script to it, it will be run before taking the snapshot. However, there are some limitations: typed arrays and function templates cannot be serialized. Also, snapshotting obviously fails if the script throws an uncaught exception. Or maybe you are running into a bug, in which case it would be great if you could file a bug :)

**Comment by subesokun** [comment]
@yangguo-chromium-org I'll try to find the piece of code that breaks the mksnapshot. Where should I file the bug (if I find any)?. Just to clarify, in case I've added e.g. React.js to the snapshot then I've to remove also the script tag which would usually load this library within my index.html or? I've noticed that if I don't remove it then I don't see any loading speed improvements which might be obvious as it tries to load it again during runtime...

**Comment by yangguo-chromium-org** [comment]
subesokun, please file V8 bugs to https://code.google.com/p/v8/issues/entry

If you include React.js again in the script tag, you will load it again, in addition to what's already in the context, from the snapshot. This may be benign, as that would be as if you loaded the script twice. But of course you don't get any benefits either.

**Comment by YuriSolovyov** [comment]
one more speed-up [just landed](https://github.com/nodejs/io.js/commit/8e53fd537193dcee3622df10e9f45028058a3477) in io.js

**Comment by ArekSredzki** [comment]
Quick question to help me with a though experiment about this:
Is the BrowserWindow require blocking/synchronous on the main thread?

**Comment by nathansobo** [comment]
@ArekSredzki If I understand your question correctly, the answer is yes. Interaction with remote requires is synchronous.

**Comment by ArekSredzki** [comment]
@nathansobo Thanks. If we promisify remote requires (I plan to implement this as an option for ipc messages anyway), then couldn't we have the atom modules perform asynchronously, and more specifically, in parallel?

Also, in general, what's the dependency hierarchy tree for these require operations like (at least on the browser side)? If there is one that is.

**Comment by nathansobo** [comment]
@ArekSredzki I think I did misunderstand slightly. All requires are synchronous. The synchronous behavior I was referring to is just that calling methods on the remote requires results in synchronous IPCs. I think we have at least one of these sync calls in our critical startup path however.

**Comment by ArekSredzki** [comment]
@nathansobo Sorry for not making myself clear, the idea that I'm exploring is whether making the IPC/renderer part of remote `require` calls asynchronous would accelerate boot up as one plugin could be initializing in the renderer thread, while another one is waiting on it's remote require to return from the sync `require()` that is running behind the scenes in the main process.

**Comment by ArekSredzki** [comment]
Essentially, the idea is to free up the renderer thread to do real computation and ensure fluid UX while the main thread deals with the only part of the require call that actually "has" to be synchronous.

**Comment by paulcbetts** [comment]
@ArekSredzki This might help but probably not as much as you think because blocking the main process will hurt render process UI performance even if render process isn't explicitly blocking on something in JS. There are lots of Chromium calls that depend on ensuring that a message has been delivered to the main process and all of those will block too while the main process is busy

**Comment by ArekSredzki** [comment]
@paulcbetts Good point.

However, the magic of node would allow the main process to continue responding to these requests while `require` is waiting on IO, would it not?

If not, it might be worth spinning up a child_process for this kind of thing.

**Comment by paulcbetts** [comment]
@ArekSredzki No, `require` is all synchronous everything :(

**Comment by ArekSredzki** [comment]
Sad days

And having a separate internal child process for handing blocking tasks?

**Comment by zcbenz** [comment]
I'm closing this now since we basically did every solution we can try. For people interested in making `require` faster, you can check the `NativeCompileCache` class in Atom:
https://github.com/atom/atom/blob/f48069c5f848b5ae624b110a1df7800e919eb142/src/native-compile-cache.js
